test/unit/org/apache/cassandra/utils/ImmutableUniqueList.java

```diff
@@ -0,0 +1,167 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.utils;
+
+import java.util.AbstractList;
+import java.util.AbstractSet;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.RandomAccess;
+
+import com.google.common.collect.Iterators;
+
+import org.agrona.collections.Object2IntHashMap;
+
+public class ImmutableUniqueList<T> extends AbstractList<T> implements RandomAccess
```

Comment:
java doesn't allow a `List` to also be a `Set` as they have method conflicts... so this `List` has the same properties of a `Set` but also maintain the insertion order dependent properties of `List`.

`LinkedHashSet` wasn't enough as the accesses pattern of `index -> Value` was cumbersome...